aten/src/ATen/Context.cpp

```diff
@@ -104,9 +104,12 @@ bool Context::setFlushDenormal(bool on) {
   return at::cpu::set_flush_denormal(on);
 }
 
+// NOTE: We also check `at::NonVariableTypeMode`, and if it's enabled we always
+// return non-Variable type in this function.
+// See NOTE [ Treating Variables as non-Variables in type dispatch ]
 TypeExtendedInterface& getType(TensorOptions options) {
   return globalContext().getType(
-            options.backend(), typeMetaToScalarType(options.dtype()), options.is_variable());
+            options.backend(), typeMetaToScalarType(options.dtype()), options.is_variable() && !at::NonVariableTypeMode::is_enabled());
```

Comment:
this call makes me a little nervous because my mental model here is that "everything is tensor, but it's Type may change depending on the thread local context."  Is that mental model wrong?

Because if that mental model is correct, what is a bit scary about this call is there's potentially conflicting information -- both the TensorOptions and the thread local state says whether something should be a variable.

I actually think this particular case is easy to resolve: this is I believe an internal function, so we can move it under at::detail.  Also, I believe this call will always not be a variable because of what is generated in variable_factories.  Can we just break out a simple change here if it makes sense?

Reply:
I think this is the mental model we should go for:

Invariant:
`Tensor is a Variable` <=> `tensor.is_variable() == true` <=> `tensor.options().is_variable() == true` <=> `TensorImpl's autograd_meta_ pointer is not null`

When a tensor's `is_variable()` is true, it means its `autograd_meta_` pointer is initialized and it's capable of recording gradients (and it will start recording gradients if we set `requires_grad` to true). Also, if a tensor's `is_variable()` is true, it means it can be cast to a Variable which provides API to access its autograd metadata (such as `grad_fn`). When a tensor's `is_variable()` is false, it means its `autograd_meta_` pointer is null.

TensorOptions's `is_variable()` decides whether to create a tensor with `autograd_meta_` pointer initialized. If a tensor is created with `options.is_variable(true)`, the tensor's `is_variable()` is guaranteed to return true, unless the `NoGradGuard` thread-local guard is active (in which case the tensor's `is_variable()` always returns false).

Whether a TensorOptions is a Variable type is solely determined by `options.is_variable()`, and we don't check the `NoGradGuard` thread-local guard there. Upstream call sites of `at::getType(options)` should set the right value for `options.is_variable()`, so that non-Variable type dispatch is used properly when needed.

Changes we need to make to achieve the above mental model:

For TensorImpl:
- We get rid of the `is_variable_` field, and `is_variable()` should check whether `autograd_meta _` pointer is null.

For TensorOptions:
- Keep `is_variable()`, and use it to decide whether we should initialize the `autograd_meta_` pointer. Add assert to make sure `requires_grad() == true` and `is_variable() == false` don't happen at the same time.

For `getType(options)`:
- `options.is_variable()` should solely decide whether a TensorOptions is a Variable type. We shouldn't check the thread-local guard here.
- Upstream call sites of `getType(options)` should set the right value for `options.is_variable()`, by passing in `options.is_variable(false)` instead of `options` when needed.

I will first open another PR to make these changes and add comments for the correct mental model, before we proceed with this PR.

Reply:
I have an alternate proposal for the semantics.

My underlying design principles are:
* Everything is a Variable (in the old sense)
* Autograd metadata being null is just an optimization for the common case, when it is not necessary

So, I don't think we should have a separate concept of `is_variable`. Instead, we should just define the situation when autograd metadata is permitted to be null. Which should be something like, autograd metadata == null implies:

* `grad_` is undefined
* `name_` is unset
* `grad_fn_` and `grad_accumulator_` are null
* there are no hooks
* does not require grad

But you are kind of in trouble for this PR, because there is some important preparatory work missing:

* `version_counter_` has to live in storage unconditionally, because we may capture non-requires-grad variables inside an autograd function, and we need a working version counter in these cases
* `pyobj_` needs to live on `TensorImpl` itself, because we *always* need to be able to convert to and from the Python representation

So it seems to me that these have to be done first, before we can apply this patch.

Reply:
@gchanan @ezyang I wrote a proposal combining our ideas:
https://github.com/pytorch/pytorch/issues/13638#issuecomment-468055428. Please feel free to review it :)

Reply:
Update: https://github.com/pytorch/pytorch/issues/13638#issuecomment-472647222 provides details on why this change is desired.

Reply:
That description doesn't say anything about `is_variable` -- is that still needed here?

Reply:
`options.is_variable()` is not always true yet - for example, in `torch::randn` defined in `torch/csrc/autograd/generated/variable_factories.h`:
```cpp
inline at::Tensor randn(at::IntArrayRef size, const at::TensorOptions & options = {}) {
  ...
  at::Tensor tensor = at::randn(size, at::TensorOptions(options).is_variable(false));
  ...
}
```

In my next PR (setting AutogradMeta only when `requires_grad=true`), I will also make `options.is_variable()` and `tensor.is_variable()` always return true, and remove the `options.is_variable()` check here.

Reply:
There is still a distinction between a Tensor and a Variable (a Tensor doesn't have AutogradMeta, while a Variable always has one). So `options.is_variable()` is still meaningful here.

Reply:
That sounds right.